lib/useOnyx.ts

```diff
@@ -0,0 +1,189 @@
+    return key.includes('_') && !key.endsWith('_');
+}
+
+function getCachedValue<TKey extends OnyxKey, TValue>(key: TKey, selector?: Selector<TKey, unknown, unknown>): CachedValue<TKey, TValue> {
+    return (Onyx.tryGetCachedValue(key, {selector}) ?? null) as CachedValue<TKey, TValue>;
+}
+
+function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnData>): UseOnyxData<TKey, TReturnData> {
+    const connectionIDRef = useRef<number | null>(null);
+    const previousKey = usePrevious(key);
+
+    /**
+     * Used to store collection objects or selected data since they aren't stored in the cache.
+     */
+    const currentDataRef = useRef<CachedValue<TKey, TReturnData> | null>(null);
+
+    const isFirstRenderRef = useRef(true);
+    const fetchStatusRef = useRef<FetchStatus>('loading');
+
+    useEffect(() => {
+        /**
+         * This condition will ensure we can only handle dynamic collection member keys.
+         */
+        if (previousKey === key || (isCollectionMemberKey(previousKey) && isCollectionMemberKey(key))) {
+            return;
+        }
+
+        throw new Error(
+            `'${previousKey}' key can't be changed to '${key}'. useOnyx() only supports dynamic keys if they are both collection member keys e.g. from 'collection_id1' to 'collection_id2'.`,
+        );
+    }, [previousKey, key]);
+
+    /**
+     * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     */
+    const getSnapshot = useCallback(() => {
+        /**
+         * Case 1 - We have a non-collection key without selector
+         *
+         * We just return the data from the Onyx cache.
+         */
+        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+            return getCachedValue(key);
+        }
+
+        /**
+         * Case 2 - We have a non-collection/collection key with/without selector
+         *
+         * Since both collection objects and selected data are not directly stored in the cache, we need to generate them with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        const newData = getCachedValue(key, options?.selector);
+        if (!deepEqual(currentDataRef.current, newData)) {
+            currentDataRef.current = newData as CachedValue<TKey, TReturnData>;
+        }
+
+        return currentDataRef.current as CachedValue<TKey, TReturnData>;
+    }, [key, options?.selector]);
```

Comment:
It's worth noting that if the selector or options are unstable references, then `getSnapshot` will be redefined with each render.

We could consider using `useRef` to stabilize the reference internally so that caller's don't have to be wary of this pitfall:

```diff
diff --git a/lib/useOnyx.ts b/lib/useOnyx.ts
index 965cf4f..9b454f9 100644
--- a/lib/useOnyx.ts
+++ b/lib/useOnyx.ts
@@ -78,6 +78,9 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
         );
     }, [previousKey, key]);
 
+    const selector = useRef(options?.selector);
+    selector.current = options?.selector;
+
     /**
      * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
      * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
@@ -89,7 +92,7 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
          *
          * We just return the data from the Onyx cache.
          */
-        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+        if (!Onyx.isCollectionKey(key) && !selector.current) {
             return getCachedValue(key);
         }
 
@@ -103,13 +106,13 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
          *
          * If they are equal, we just return the previous internal data.
          */
-        const newData = getCachedValue(key, options?.selector);
+        const newData = getCachedValue(key, selector.current);
         if (!deepEqual(currentDataRef.current, newData)) {
             currentDataRef.current = newData as CachedValue<TKey, TReturnData>;
         }
 
         return currentDataRef.current as CachedValue<TKey, TReturnData>;
-    }, [key, options?.selector]);
+    }, [key, selector]);
 
     const subscribe = useCallback(
         (onStoreChange: () => void) => {
```

I've also seen some people wrap this pattern with a simple `useLiveRef` utility hook:

```ts
function useLiveRef<T>(value: T) {
  const ref = useRef(value);
  ref.current = value;
  return ref;
}

...

const selector = useLiveRef(options?.selector);
```

That would allow us to safely use this hook like so:

```ts
const report = useOnyx(ONYXKEYS.COLLECTION.REPORT, {
    selector: (reports) => reports[reportID],
});
```

Instead of having to do some verbose memoization:

```ts
const selector = useCallback((reports) => reports[reportID], [reportID]);
const options = useMemo(() => ({selector}), [selector]);
const report = useOnyx(ONYXKEYS.COLLECTION.REPORT, options);
```

Reply:
That makes sense to me, thanks for the suggestion!